### DISCUSSION TITLE
Continuous collision detection: body jerking

### DISCUSSION THREAD

**Discussion by Krogenit** [dyn4j/dyn4j#278] [Q&A]
I have issue with body jerking when my bullets collides with other bodys. 

Issue example: https://youtu.be/wDVA0SxsE4w

I know that the CCD translates the bodies to the time of the collision. I did workaround with saving transforms of affected bodies and then restore thier after World.step(). I think it affects bad for collision precision.

Workaround example: https://youtu.be/O9pLqLha2UM

``` java
listener.clear();
world.step(1);
listener.restoreTransforms();
```

``` java
public class ContactListener extends TimeOfImpactListenerAdapter<Body> {
    private final THashSet<GameObject> affectedBodies = new THashSet<>();

    @Override
    public boolean collision(Body body1, Body body2, TimeOfImpact toi) {
        saveTransform(((GameObject) body1.getUserData()), body1);
        saveTransform(((GameObject) body2.getUserData()), body2);
        return true;
    }

    private void saveTransform(GameObject gameObject, Body body) {
        gameObject.saveTransform(body.getTransform());
        affectedBodies.add(gameObject);
    }

    public void restoreTransforms() {
        affectedBodies.forEach(object -> {
            object.restoreTransform();
            return true;
        });
    }

    public void clear() {
        affectedBodies.clear();
    }
}
```

Is there a better way to solve that?

**Comment by wnbittle** [comment]
Have you tried disabling CCD entirely?  From the videos, it doesn't seem like the bullets are travelling that fast so maybe you don't need CCD.  If you find that disabling CCD causes missed collisions, you can try decreasing the `Settings.getStepFrequency` to detect collisions more frequently - naturally this could have a performance impact, but it can work for small to medium sized simulations.

Another option might be to handle the CCD event manually with your own resolution step.  To do that, return `false` from the `collision(Body, Body, TimeOfImpact)` method to tell dyn4j to NOT translate the bodies to the time of collision.  Then, you could handle it in a different way depending on what you need.  For example, you could translate ONLY the bullet to the time of impact.

The game and gameplay look really nice btw!

**Reply by Krogenit** [reply]
Thank you for your reply!

I have tried without CCD, but it wasn't good. I agreed that bullets aren't fast but they were passing through ship geometry a bit and collision was inside ship. It also happens with saving & restoring transforms. I set 1 / 60 step frequency and think this is enough for this type of game and also for optimization purposes for future large battles that I planning.

For example with ```ContinuousDetectionMode.NONE``` or saving & restoring transforms it happens often:

![hole](https://user-images.githubusercontent.com/37593387/226108673-fc7f5629-3994-48a7-a9f1-f2e24d1ae799.PNG)

I tried your suggestion with translating only bullets to the time of collision, but there odd behaviour happens, bullets stick to ship: https://youtu.be/pRFiHB2JELI

Think need more time for it. Anyways thanks for the help!
